### PR TITLE
structs: namespace port validation by host_network

### LINKED ISSUE #8891: Problems with new Multi Interface Networking Feature with same ports to different interfaces.
### Nomad version
Nomad v0.12.4

### Operating system and Environment details
Linux edge0 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

### Issue
I would expect the following configuration to work, however Nomad complains of a collision. This is actually not true, but nomad gets confused as previously without multi interface networking this would have collided.

**Nomad Job**
```
    network {
      mbits = 10
      mode = "bridge"
      port "http" {
        static = 80
        host_network = "public"
      }
      port "https" {
        static = 443
        host_network = "public"
      }
      port "http" {
        static = 80
        host_network = "localhost"
      }
      port "https" {
        static = 443
        host_network = "localhost"
      }
    }
```

**Nomad Config**
```
{
....
  "client": {
    "enabled": true,
    "host_network": {
      "private": {
        "interface": "eth1",
      },
      "public": {
        "interface": "eth0",
      },
      "localhost": {
        "interface": "lo"
      }
    }
....
}
```

When attempting the above job, which I would expect to work however it shows:
`error parsing 'job': group: network, ports -> found a port label collision: http`

### LINKED ISSUE #9477: Static ports on different host networks should not conflict
### Nomad version
`Nomad v0.12.7 (6147cb578794cb2d0c35d68fe1791728a09bb081)`

### Operating system and Environment details
Debian 10

### Issue
Static ports in jobs with the same number, but with different `host_network` settings are treated as conflicting.  They should be able to run on the same host as they would bind to distinct IP addresses.  It appears that this persists beyond the "in the same job" scenario, as launching different jobs on the same host with the same network configuration will fail due to resource exhaustion.  Should I submit a separate issue for that?

#### ensure the nomad.hcl has some host_networks defined 
```
client {
  enabled       = true
  network_speed = 1000
  host_network "net1" { 
    cidr = "10.9.9.1/32"
  }
  host_network "net2" { 
    cidr = "10.9.9.2/32"
  }
}
```
#### bring up a couple new dummy interfaces with our IPs
```
# ip link add name nomad1 type dummy
# ip link add name nomad2 type dummy

# ip addr add 10.9.9.1/32 dev nomad1
# ip addr add 10.9.9.2/32 dev nomad2

# ip link set up dev nomad1
# ip link set up dev nomad2
```
#### restart nomad, try to run our job
```
# systemctl restart nomad
# nomad job plan static-port-test.hcl
Error during plan: Unexpected response code: 500 (1 error occurred:
        * Task group group1 validation failed: 1 error occurred:
        * Task group network validation failed: 1 error occurred:
        * Static port 32767 already reserved by taskgroup network:net1_port_32767
```

### Job file (if appropriate)
```
job "static-port-test" {
  region = "global"
  datacenters = ["dc1"]
  type = "batch"

  group "group1" {

    network {
      port "net1_port_32767" {
        static = 32767
        to = 32767
        host_network = "net1"
      }
      port "net2_port_32767" {
        static = 32767
        to = 32767
        host_network = "net2"
      }
    }

    task "net1_task" {
      driver = "docker"

      config {
        image = "alpine/socat:latest"
        args = [ "tcp-listen:32767", "file:/etc/hostname" ]
        ports = [ "net1_port_32767" ]
      }

      resources {
        cpu    = 20 # MHz
        memory = 16 # MB
      }
    }

    task "net2_task" {
      driver = "docker"

      config {
        image = "alpine/socat:latest"
        args = [ "tcp-listen:32767", "file:/etc/hostname" ]
        ports = [ "net2_port_32767" ]
      }

      resources {
        cpu    = 20 # MHz
        memory = 16 # MB
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by nickethier**
Previously static port validation did not take host_networks into account. This meant Nomad thought two static ports with the same value but separate host_networks conflicted. The scheduler already tracks port allocations by the assigned HostIP so the issue is solely wrongful validation.

fixes #9477 
fixes #8891

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,10 +13,11 @@ BUG FIXES:
...
  * driver/qemu: Fixed a bug where network namespaces were not supported for QEMU workloads [[GH-9861](https://github.com/hashicorp/nomad/pull/9861)]
+ * nomad/structs: Fixed a bug where static ports with the same value but different `host_network` were invalid [[GH-9946](https://github.com/hashicorp/nomad/issues/9946)]
  * scheduler: Fixed a bug where shared ports were not persisted during inplace updates for service jobs. [[GH-9830](https://github.com/hashicorp/nomad/issues/9830)]
  * scheduler: Fixed a bug where job statuses and summaries where duplicated and miscalculated when registering a job. [[GH-9768](https://github.com/hashicorp/nomad/issues/9768)]
+ * scheduler: Fixed a bug that caused the scheduler not to detect changes for `host_network` port field. [[GH-9973](https://github.com/hashicorp/nomad/issues/9973)]
  * scheduler (Enterprise): Fixed a bug where the deprecated network `mbits` field was being considered as part of quota enforcement. [[GH-9920](https://github.com/hashicorp/nomad/issues/9920)]
- * volumes: Fixed a bug where volume diffs were not displayed in the output of `nomad plan`. [[GH-9973](https://github.com/hashicorp/nomad/issues/9973)]
```
Comment (shoenig) [inline CHANGELOG.md:19]: Was this line intended to be removed?

**Review by shoenig — APPROVED** [review]
LGTM

File: CHANGELOG.md (hunk 1 above)
Comment (nickethier) [reply CHANGELOG.md:19]: Hmm no that might have came from a rebase. Fixing up!

**Comment by mr-karan** [comment]
Eagerly waiting for this to get merged :D

**Pull request merged**